### ISSUE TITLE
handle libvlc 3.0 or under API versioning

### ISSUE THREAD

**Issue by mfkl** [mfkl/LibVLCSharp-readonly#8]
In libvlcpp, it's currently done with an [#if def](https://code.videolan.org/videolan/libvlcpp/blob/master/vlcpp/Instance.hpp#L512). 

Would it be a good idea to check the [LIBVLC_API](https://www.videolan.org/developers/vlc/doc/doxygen/html/group__libvlc.html#gaee48476813eb6eaa37b22df85099d76e) C macro value at runtime of the used `.dll` before making libvlc >= 3.0 API calls? 

Guess that has a runtime cost, but not sure how else to dynamically support `dlls` with different public APIs with the same code.

**Comment by mfkl** [comment]
added https://github.com/mfkl/libvlcsharp/blob/8f30c5331199f705ea247686349034de08cf3d35/LibVlcSharp/Manual/Instance.cs#L258

**Comment by mfkl** [comment]
This is implemented, however:
- Do we want to check at compile time (and possibly strip methods that are over the supported API version) ? We could do something with [Fody](https://github.com/Fody/Fody) which I've used in the past to hook into the build pipeline.
- Do we want a runtime check? One suggestion from @jeremyVignelles would be to use an attribute, such as `[RequiresLibvlcVersion(3)]`. That would be pretty neat, too.
- Do we want both a runtime and a compile check?

For the record libvlcpp uses a compile check only (via C++ compilation directive).

**Comment by mfkl** [comment]
- if we're to be able to compile `libvlcsharp` *without* referencing `libvlc.dll`, a compile time check (i.e: strip) is not possible. Probably should go runtime only.
- .NET custom attributes don't allow running code before method is executed out of the box. We need to use a tool to detect our attribute and inject the code (that will perform the check at runtime).
- I know Fody. Roslyn analyzer can do the job as well though I'm not familiar with them.

**Comment by jeremyVignelles** [comment]
Fody solves another kind of problems. If I remember correctly, fody is about modifying an assembly at build time by applying transformations to it. .net analyzers, on the other hand, can be embedded into a NuGet package and is automatically inserted in the project referencing the NuGet package, allowing it to check potential usage issues.
I don't really know how to apply that though. we probably need a define in the code to tell the version we are intending to use?

**Comment by mfkl** [comment]
> Fody solves another kind of problems. If I remember correctly, fody is about modifying an assembly at build time by applying transformations to it.

Yeah, and that's exactly what we want the way I see it. Goal is to check the `dll` version before each 3.0 (or even 2.2) calls. We don't want to write that code everytime or have a global static class that performs the check. With just an attribute, Fody can inject the code we need and we keep our code clean.

> .net analyzers, on the other hand, can be embedded into a NuGet package and is automatically inserted in the project referencing the NuGet package, allowing it to check potential usage issues.

That would be good for warning the user at compile time (or even before, as soon as they reference the `.dll`). But then we still need to throw if the user makes the call when he shouldn't and that requires implementing (another) codefix provider for code generation at build time.

**Comment by jeremyVignelles** [comment]
Agreed.
Is Fody really required ? Couldn't we do that with cppsharp directly?
I'm still thinking about it. Runtime check is required to be sure, but this comes with a negative impact, especially for frequently called methods.

**Comment by mfkl** [comment]
This Fody addin seems to do the job nicely https://github.com/Fody/MethodDecorator

**Comment by mfkl** [comment]
Ended up using https://github.com/Capgemini/Cauldron/tree/master/Fody/Cauldron.Interception.Fody for netstandard 2.0 support.
Need to add support for:
- ctor interception,
- not auto-property interception.
